### PR TITLE
[Tests-Only] Refactor assert method in file BlogPostsControllerTest.php 

### PR DESCRIPTION
**Pull request by swoichha**
### Related Issues 
https://github.com/foodcoopshop/foodcoopshop/issues/404

### Description
- Refactored assert method in test  `BlogPostsControllerTest.php` and used assert method of `IntegrationTestTrait`.
- Renamed method `loginAsCustomer` to `loginAsCustomerWithHttpClient` in AppCakeTestCase.php

### CODE REVIEW

File: tests/TestCase/src/Controller/BlogPostsControllerTest.php [hunk 1]
```diff
@@ -29,41 +32,43 @@ public function setUp(): void
 
     public function testBlogPostDetailOnlinePublicLoggedOut()
     {
-        $this->httpClient->get($this->Slug->getBlogPostDetail(2, 'Demo Blog Artikel'));
-        $this->assert200OkHeader();
+        $_SERVER['REQUEST_URI'] = $this->Slug->getBlogPostDetail(2, 'Demo Blog Artikel');
```
Comment (mrothauer) [inline tests/TestCase/src/Controller/BlogPostsControllerTest.php:35]: $_SERVER['REQUEST_URI'] should not be set in test case. Is there a better solution to stop MenuHelper complaining?

File: tests/TestCase/src/Controller/BlogPostsControllerTest.php [hunk 2]
```diff
@@ -29,41 +32,43 @@ public function setUp(): void
...
 
     public function testBlogPostDetailOnlinePrivateLoggedOut()
     {
         $blogPostId = 2;
         $this->changeBlogPost($blogPostId, 1);
-        $this->httpClient->followOneRedirectForNextRequest();
-        $this->httpClient->get($this->Slug->getBlogPostDetail($blogPostId, 'Demo Blog Artikel'));
-        $this->assertAccessDeniedWithRedirectToLoginForm();
+        $this->get($this->Slug->getBlogPostDetail($blogPostId, 'Demo Blog Artikel'));
+        $this->assertRedirectContains('/anmelden?redirect=%2Faktuelles%2F2-Demo-Blog-Artikel');
```
Comment (mrothauer) [inline tests/TestCase/src/Controller/BlogPostsControllerTest.php:53]: Please use SlugHelper->getLogin() here (it accepts a redirect param)

**Comment by mrothauer** [comment]
Thank's a lot - good work! Please let me know when you're done with your work and it's my turn to merge the PR.

File: tests/TestCase/src/Controller/BlogPostsControllerTest.php (hunk 1 above)
Comment (individual-it) [reply tests/TestCase/src/Controller/BlogPostsControllerTest.php:35]: I see multiple other options:
1. define `$_SERVER['REQUEST_URI']` in `phpunit.xml` see `<server>` element in https://phpunit.readthedocs.io/en/9.0/configuration.html#the-php-element
2. in the MenuHelper check if code is run from phpunit similar to `header.php` https://github.com/foodcoopshop/foodcoopshop/pull/529/files#diff-dd7e8b4e9539e4296688523beebb8916R22
3. hide the warnings

I think the question is what we want to test, if the MenuHelper is part of the test we should set it to something realistic (mean leave it as it is), if not we can go for one of the other possibilities

Comment (mrothauer) [reply tests/TestCase/src/Controller/BlogPostsControllerTest.php:35]: I think option 1 is good for us. MenuHelper is not part of the tests. Thanks!

File: src/View/Helper/MenuHelper.php [hunk 3]
```diff
@@ -103,7 +103,6 @@ private function renderMenuElement($slug, $name, $style = '', $class = [], $font
         }
         if ($slug != '/' && preg_match('`' . preg_quote($slug) . '`', $_SERVER['REQUEST_URI'])) {
             $applyActiveClass = true;
-
```
Comment (mrothauer) [inline src/View/Helper/MenuHelper.php:106]: Please remove this line so the MenuHelper remains untouched.

File: tests/TestCase/src/Controller/BlogPostsControllerTest.php [hunk 4]
```diff
@@ -29,41 +32,43 @@ public function setUp(): void
...
 
     public function testBlogPostDetailOnlinePrivateLoggedOut()
     {
         $blogPostId = 2;
         $this->changeBlogPost($blogPostId, 1);
-        $this->httpClient->followOneRedirectForNextRequest();
-        $this->httpClient->get($this->Slug->getBlogPostDetail($blogPostId, 'Demo Blog Artikel'));
-        $this->assertAccessDeniedWithRedirectToLoginForm();
+        $this->get($this->Slug->getBlogPostDetail($blogPostId, 'Demo Blog Artikel'));
+        $this->Slug->getLogin($blogPostId, 'Demo Blog Artikel');
```
Comment (mrothauer) [inline tests/TestCase/src/Controller/BlogPostsControllerTest.php:53]: You need to assert the result of Slug->getLogin()
eg. `$this->assertRedirectContains($this->Slug->getLogin($blogPostId, 'Demo Blog Artikel'));`
eventually you nedd to move urlencode from here to getLogin() https://github.com/foodcoopshop/foodcoopshop/blob/81bd838dd1071efeb54b9756583302ba06d0be07/templates/Customers/login.php#L32 into getLogin()

File: tests/TestCase/src/Controller/BlogPostsControllerTest.php [hunk 5]
```diff
@@ -29,41 +32,43 @@ public function setUp(): void
...
     }
 
     public function testBlogPostDetailOnlinePrivateLoggedIn()
     {
         $this->loginAsCustomer();
         $blogPostId = 2;
         $this->changeBlogPost($blogPostId, 1);
-        $this->httpClient->get($this->Slug->getBlogPostDetail($blogPostId, 'Demo Blog Artikel'));
-        $this->assert200OkHeader();
+        $_SERVER['REQUEST_URI'] = $this->Slug->getBlogPostDetail($blogPostId, 'Demo Blog Artikel');
```
Comment (mrothauer) [inline tests/TestCase/src/Controller/BlogPostsControllerTest.php:62]: Please use option 1 (define server-element in phpunit.dist.xml)
https://github.com/foodcoopshop/foodcoopshop/pull/529#discussion_r431565459

File: tests/TestCase/src/Controller/BlogPostsControllerTest.php (hunk 4 above)
Comment (swoichha) [reply tests/TestCase/src/Controller/BlogPostsControllerTest.php:53]: It should be something like `$this->assertRedirectContains($this->Slug->getLogin('%2Faktuelles%2F'.$blogPostId.'-Demo-Blog-Artikel'));`. What do you think?

Comment (mrothauer) [reply tests/TestCase/src/Controller/BlogPostsControllerTest.php:53]: The result should be the same as this:
`'/anmelden?redirect=%2Faktuelles%2F2-Demo-Blog-Artikel'`
And please also use SlugHelper->getBlogPostDetail and do not manually build the url.

**Comment by mrothauer** [comment]
thanks for your work. I had to add some commits. next time I push to the fork to have them in the same PR, I messed that up...
e2427bbae4ae67dc5526fff09d133d306fac41d7
1d9e14a1e6c71aa9e8e2923d1d98c11a19cb3e71
7a11c0f720ddff044af25bdaef82fd81bc45779f
32a2221f57ec14848b69ec4bfd227f411eff5c46

**Pull request merged**